src/types/pick.test.ts

```diff
@@ -0,0 +1,219 @@
+const TestType = object('TestType', {
+    a: string,
+    b: boolean,
+    c: number,
+});
+
+testTypeImpl({
+    name: `Pick<TestType, 'b' | 'c'>`,
+    type: pick(TestType, ['b', 'c']),
+    basicType: 'object',
+    validValues: [
+        { b: true, c: 5 },
+        { b: false, c: -2 },
+    ],
+    invalidValues: [
+        [{ b: 'blah', c: 3 }, `error in [Pick<TestType, 'b' | 'c'>] at <b>: expected a boolean, got a string ("blah")`],
+        [{ c: 3 }, `error in [Pick<TestType, 'b' | 'c'>]: missing property <b> [boolean], got: { c: 3 }`],
+    ],
+    validConversions: [
+        [
+            { a: 'drop', b: true, c: 2 },
+            { b: true, c: 2 },
+        ],
+    ],
+});
+
+type PartialTestType = The<typeof PartialTestType>;
+const PartialTestType = partial('PartialTestType', {
+    d: string,
+    e: number,
+});
+testTypeImpl({
+    name: `Pick<PartialTestType, 'd'>`,
+    type: pick(PartialTestType, ['d']),
+    basicType: 'object',
+    validValues: [{ d: 'blah' }, {}],
+    invalidValues: [[{ d: undefined }, `error in [Pick<PartialTestType, 'd'>] at <d>: expected a string, got an undefined`]],
+    validConversions: [
+        [{ b: true, c: 3 }, {}],
+        [{ d: 'hns', e: 1 }, { d: 'hns' }],
+    ],
+});
+
+type IntersectionTestType = The<typeof IntersectionTestType>;
+const IntersectionTestType = intersection('IntersectionTestType', [TestType, PartialTestType]);
+testTypeImpl({
+    name: `Pick<IntersectionTestType, 'a' | 'd'>`,
+    type: pick(IntersectionTestType, ['a', 'd']),
+    basicType: 'object',
+    validValues: [{ a: 'blah', d: 'dah' }, { a: 'blah2' }, { a: 'hts', d: '' }],
+    invalidValues: [
+        [{ d: 'dah' }, `error in [Pick<IntersectionTestType, 'a' | 'd'>]: missing property <a> [string], got: { d: "dah" }`],
+        [{ a: 'blah2', d: undefined }, `error in [Pick<IntersectionTestType, 'a' | 'd'>] at <d>: expected a string, got an undefined`],
+    ],
+    validConversions: [
+        [
+            { a: 'drop', b: true, c: 2, d: 'tnsh', e: 2 },
+            { a: 'drop', d: 'tnsh' },
+        ],
+        [{ a: 'tns' }, { a: 'tns' }],
+    ],
+});
+
+type Overlap = The<typeof Overlap>;
+const Overlap = object('Overlap', {
+    a: string,
+    b: boolean,
+    d: boolean,
+});
+type UnionTestType = The<typeof UnionTestType>;
+const UnionTestType = union('UnionTestType', [TestType, Overlap]);
+// const pickunion = pick(UnionTestType, ['a']);
+testTypeImpl({
+    name: `Pick<UnionTestType, 'a'>`,
+    type: pick(UnionTestType, ['a']),
+    basicType: 'object',
+    validValues: [{ a: 'str' }, { a: '' }],
+    invalidValues: [
+        [{}, `error in [Pick<UnionTestType, 'a'>]: missing property <a> [string | string], got: {}`],
+        [{ a: true }, `error in [Pick<UnionTestType, 'a'>] at <a>: expected a string, got a boolean (true)`],
+    ],
+    validConversions: [
+        [{ a: 'Overlap', b: true }, { a: 'Overlap' }],
+        [{ a: 'TestType', b: false, c: 12 }, { a: 'TestType' }],
+    ],
+});
```

Comment:
Ok, this is only one of the interesting cases with unions (overlap with the same property type) there are other that need to be tested (and some do not work as expected yet):
- [x] `Pick<{ a: string, b: boolean} | { a: string, b: boolean }, 'a'>`
- [ ] `Pick<{ a: string } | { a: boolean }, 'a'>` should accept both a `string` and a `boolean` for `a`
- [ ] But even more interesting, a discriminated union:
    ```ts
    Pick<
        | { t: 'string', v: string, obsolete: string }
        | { t: 'number', v: number, otherStuffBeGone: unknown },
        't' | 'v'
    >
    ```
    Here we have a choice, but I believe we should distribute over the elements of the union, see [this playground link](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgEWAZxlYAjArjAUwBMAFYAYwGs4BfOAMyghDgHIYBPMAgWnoIysA3ACgRnbnACyHAKoA7YBHlwAvCLia4AH0RwYALjYYs8gOasANHABuRk8HPWI2NBAA2BQvcyOztDS1dJEM2eVwQbAIoK1sjcMjo5xgAC2iAZRhcenoAIQIAcWUCI1x5SnkIAHcVGlFxLgIUAnoAQ1x3GHyU1pslKDU4MioAHhkFJXlrdlYdNhtWAD56gHoVuAAVFPQ4HYhKAxEAChCjVgSomOs7OABGACYAZloASjg0Vph0emABZraOl0CD0+tBRCd9GcHOZYjdWK13r4Ya93p9vr80P92p1ur1+qt1rl8HBWvRCANUk1htRMK15Gh6NAQNZWu53HAwExuLAMSSoE0AERlSbAABexAFu3kRGAfSIuFZ7g41ncEDQfn0aTgMrQ5CwIEcn2IcGFyjgURB-QAdMdTmEIpdYWcICk4JUAISzGhvD5fBm85AtbFAy1gsQiNZwACS5oI7l+NiaqU+u0xrJAavg8gIiYGVR68A4EFwcCqxfcRFLdJgVrgsnV5jg8dcTeAlCaAAMJLx+BgOyTpXAu2geDAIB4op1+1UmmYvCTtehfHhCJWwBQaVA6QymYcGpIAOrV4i40EDVQoJdYFfEaljOSKZTTGCzXSsBbLMQQ0LnB3RJ13E8KK+uifxHvIq6nvitqQsYSIWNcZwItCFjAWi-pgceRBQWGEbrAAAjAI4EAAHtw5AwDw0RMAMUTkO0aBNOBkHAni0CpiSZitI4C46nqoCGquJqPioME-hc-6IWwLpuhAnpoX6PyYRBJ6sWeoiRoRxFkQQFFUVANFiVC8EAQ8zzeqiim8sxqmhlAohAA) for more info.
- [ ] And what should happen in this case?
    ```ts
    Pick<
        | { t: 'string', v: string, obsolete: string }
        | { t: 'number', v: number, otherStuffBeGone: unknown }
        | { t: 'none' }, // this variant has no `v` property, are we allowed to pick 'v'?
        't' | 'v'
    >
    ```
    See [this playground link](https://www.typescriptlang.org/play/?ssl=7&ssc=1&pln=6&pc=22#code/JYWwDg9gTgLgBAbzgEWAZxlYAjArjAUwBMAFYAYwGs4BfOAMyghDgHIYBPMAgWnoIysA3ACgRnbnACyHAKoA7YBHlwAvCLia4AH0RwYALjYYs8gOasANHABuRk8HPWI2NBAA2BQvcyOztDS1dJEM2eVwQbAIoK1sjcMjo5xgAC2iAZRhcenoAIQIAcWUCI1x5SnkIAHcVGkDNYP0jVkr5AlZaITgAem79FPRbAEMsIfl4FKG0OEq4AAMbObgwJm5YDmsRgjgq7aH3d2rifQhlimpWG1YAfjFeuAABGDQeAgAPbnIYV6gmKBmIPA0LgwJBYMdsBx+ts8MB3DBHHAyFR9FAxmh6NAQNZdnByGM4Mp3FCwOdlqtogiBP0hvB3uh4Ij9u5hqNxmgAHTiLjbZAEehDXDw-KTGxKf6qJHnAA8MgUSnk1nYHV0l1YAD5RCJ7rl8HBgIzpsDQdBCEQ4JD+sAoOawCNOGcURAwJzuZIAOpjM0ioZi6BqFAMrB4M3IyiyuSKZRKmAqthXTViAAUIWaCSiMWsdjgAEYAEwAZloAEo4Ghaeh6MBqZ7xsQfX6oKIU01jL5zLFs6whmX2xYS2WKxjq9Na96CKLxc3U2Fih0aKXywjhzWvfWJ76p3c+k8Xu9Pt9on8RC3Qi0IhnO80ICkAQBCeeLodV1d1ogNrf3XevD4EL4-Y9T2aBwOyzIx8yLBdB2XF9RzXd8N0bUQv2eH8DwA6AjAAeWwAArP94HcA1on2OAQCGKEiShNBPmAegoQqaoVBWZ1KRHTZ5HNNU4CICBqUqOk3gZfUVAkbZWDDaUZwAIlaAhpK6GhrGkmBpPVE8ZxaOcwNzQsByXSsRzgMd10naBRCAA)